### ISSUE TITLE
HPX application fail during finalization

### ISSUE THREAD

**Issue by parsa** [STEllAR-GROUP/hpx#1467]
GCC 4.9.0
SuperMIC
Release, RelWithDebInfo
MPI: MVAPICH 2
Boost 1.57.0
```
$ mpirun -f actual.nodes $WORK/build/gcc-release/bin/hello_world -v
WARNING: Error in initializing MVAPICH2 ptmalloc library.Continuing without InfiniBand registration cache support.

HPX - High Performance ParalleX
A general purpose parallel C++ runtime system for distributed applications
of any scale.

Copyright (c) 2007-2015, The STE||AR Group,
http://stellar-group.org, email:hpx-users@stellar.cct.lsu.edu

Distributed under the Boost Software License, Version 1.0. (See accompanying
file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)

Versions:
  HPX: V0.9.11-trunk (AGAS: V3.0), Git: 9fab4ab8e5
  Boost: V1.55.0
  Hwloc: V1.9.0
  MPI: MPICH V3.1, MPI V3.0

Build:
  Type: release
  Date: Apr 18 2015 22:08:38
  Platform: linux
  Compiler: GNU C++ version 4.9.0
  Standard Library: GNU libstdc++ version 20140422
  Allocator: tcmalloc


HPX - High Performance ParalleX
A general purpose parallel C++ runtime system for distributed applications
of any scale.

Copyright (c) 2007-2015, The STE||AR Group,
http://stellar-group.org, email:hpx-users@stellar.cct.lsu.edu

Distributed under the Boost Software License, Version 1.0. (See accompanying
file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)

Versions:
  HPX: V0.9.11-trunk (AGAS: V3.0), Git: 9fab4ab8e5
  Boost: V1.55.0
  Hwloc: V1.9.0
  MPI: MPICH V3.1, MPI V3.0

Build:
  Type: release
  Date: Apr 18 2015 22:08:38
  Platform: linux
  Compiler: GNU C++ version 4.9.0
  Standard Library: GNU libstdc++ version 20140422
  Allocator: tcmalloc


hello world from OS-thread 0 on locality 0
hello world from OS-thread 0 on locality 1
{stack-trace}: 12 frames:
0x2acffea2e316  : hpx::termination_handler(int) + 0x126 in /work/parsa/build/gcc-release/lib/libhpx.so.0
0x307240f710    : ??? + 0x307240f710 in /lib64/libpthread.so.0
0x2acffeaa85e5  : hpx::naming::detail::intrusive_ptr_release(hpx::naming::detail::id_type_impl*) + 0x5 in /work/parsa/build/gcc-release/lib/libhpx.so.0
0x2acffed98eba  : hpx::util::reinit_destruct() + 0x4a in /work/parsa/build/gcc-release/lib/libhpx.so.0
0x2acffea2f121  : hpx::runtime::~runtime() + 0x1d1 in /work/parsa/build/gcc-release/lib/libhpx.so.0
0x2acffe9e21e5  : hpx::runtime_impl<hpx::threads::policies::local_priority_queue_scheduler<boost::mutex, hpx::threads::policies::lockfree_fifo, hpx::threads::policies::lockfree_fifo, hpx::threads::policies::lockfree_lifo>, hpx::threads::policies::callback_notifier>::~runtime_impl() + 0x165 in /work/parsa/build/gcc-release/lib/libhpx.so.0
0x2acffe9e2329  : hpx::runtime_impl<hpx::threads::policies::local_priority_queue_scheduler<boost::mutex, hpx::threads::policies::lockfree_fifo, hpx::threads::policies::lockfree_fifo, hpx::threads::policies::lockfree_lifo>, hpx::threads::policies::callback_notifier>::~runtime_impl() + 0x9 in /work/parsa/build/gcc-release/lib/libhpx.so.0
0x2acffea0a0dc  : ??? + 0x2acffea0a0dc in /work/parsa/build/gcc-release/lib/libhpx.so.0
0x2acffea0b60d  : hpx::detail::run_or_start(hpx::util::function<int (boost::program_options::variables_map&), void, void> const&, boost::program_options::options_description const&, int, char**, std::vector<std::string, std::allocator<std::string> > const&, hpx::util::function<void (), void, void> const&, hpx::util::function<void (), void, void> const&, hpx::runtime_mode, bool) + 0x29d in /work/parsa/build/gcc-release/lib/libhpx.so.0
0x440c46        : ??? + 0x440c46 in /work/parsa/build/gcc-release/bin/hello_world
0x44109a        : ??? + 0x44109a in /work/parsa/build/gcc-release/bin/hello_world
0x3071c1ed5d    : __libc_start_main + 0xfd in /lib64/libc.so.6
{what}: Segmentation fault
{config}:
  HPX_HAVE_NATIVE_TLS=ON
  HPX_HAVE_STACKTRACES=ON
  HPX_HAVE_COMPRESSION_BZIP2=OFF
  HPX_HAVE_COMPRESSION_SNAPPY=OFF
  HPX_HAVE_COMPRESSION_ZLIB=OFF
  HPX_HAVE_PARCEL_COALESCING=ON
  HPX_PARCELPORT_TCP=OFF
  HPX_PARCELPORT_MPI=ON (MPICH V3.1, MPI V3.0)
  HPX_PARCELPORT_IPC=OFF
  HPX_PARCELPORT_IBVERBS=OFF
  HPX_HAVE_VERIFY_LOCKS=OFF
  HPX_HAVE_HWLOC=ON
  HPX_HAVE_ITTNOTIFY=OFF
  HPX_RUN_MAIN_EVERYWHERE=OFF
  HPX_LIMIT=5
  HPX_PARCEL_MAX_CONNECTIONS=512
  HPX_PARCEL_MAX_CONNECTIONS_PER_LOCALITY=4
  HPX_INITIAL_AGAS_LOCAL_CACHE_SIZE=256
  HPX_AGAS_LOCAL_CACHE_SIZE_PER_THREAD=32
  HPX_MALLOC=tcmalloc
  HPX_PREFIX (configured)=/work/parsa/build/gcc-release
  HPX_PREFIX=/work/parsa/build/gcc-release
{version}: V0.9.11-trunk (AGAS: V3.0), Git: 9fab4ab8e5
{boost}: V1.55.0
{build-type}: release
{date}: Apr 18 2015 22:08:38
{platform}: linux
{compiler}: GNU C++ version 4.9.0
{stdlib}: GNU libstdc++ version 20140422
[smic103:mpi_rank_0][error_sighandler] Caught error: Aborted (signal 6)

===================================================================================
=   BAD TERMINATION OF ONE OF YOUR APPLICATION PROCESSES
=   PID 110928 RUNNING AT smic103
=   EXIT CODE: 6
=   CLEANING UP REMAINING PROCESSES
=   YOU CAN IGNORE THE BELOW CLEANUP MESSAGES
===================================================================================
YOUR APPLICATION TERMINATED WITH THE EXIT STRING: Aborted (signal 6)
This typically refers to a problem with your application.
Please see the FAQ page for debugging suggestions
```

**Comment by hkaiser** [comment]
Try using `HPX_MALLOC=custom`, this might avoid the initialization issue.

**Comment by parsa** [comment]
Boost, hwloc recompiled, allocator set to custom (ptmalloc is being used now). Fixed.

**Comment by parsa** [comment]
My mistake, it didn't happen in debug mode I though it's fixed.
GCC 4.9.0, Boost 1.57.0, SuperMIC
Stack trace form relwithdebinfo:
```
{stack-trace}: 12 frames:
0x2aff4af4c6f6  : hpx::termination_handler(int) + 0x126 in /worka/work/parsa/gcc49/hpx-relwithdebinfo/lib/libhpx.so.0
0x3cb9a0f710    : ??? + 0x3cb9a0f710 in /lib64/libpthread.so.0
0x2aff4d4eebaf  : _int_free + 0x5f in /usr/local/packages/mvapich2/2.0/INTEL-14.0.2/lib/libmpich.so.12
0x2aff4d4f09fc  : free + 0xcc in /usr/local/packages/mvapich2/2.0/INTEL-14.0.2/lib/libmpich.so.12
0x2aff4b2ad0ea  : hpx::util::reinit_destruct() + 0x4a in /worka/work/parsa/gcc49/hpx-relwithdebinfo/lib/libhpx.so.0
0x2aff4af4f6a1  : hpx::runtime::~runtime() + 0x261 in /worka/work/parsa/gcc49/hpx-relwithdebinfo/lib/libhpx.so.0
0x2aff4af04c0f  : hpx::runtime_impl<hpx::threads::policies::local_priority_queue_scheduler<boost::mutex, hpx::threads::policies::lockfree_fifo, hpx::threads::policies::lockfree_fifo, hpx::threads::policies::lockfree_lifo>, hpx::threads::policies::callback_notifier>::~runtime_impl() + 0x14f in /worka/work/parsa/gcc49/hpx-relwithdebinfo/lib/libhpx.so.0
0x2aff4af04d59  : hpx::runtime_impl<hpx::threads::policies::local_priority_queue_scheduler<boost::mutex, hpx::threads::policies::lockfree_fifo, hpx::threads::policies::lockfree_fifo, hpx::threads::policies::lockfree_lifo>, hpx::threads::policies::callback_notifier>::~runtime_impl() + 0x9 in /worka/work/parsa/gcc49/hpx-relwithdebinfo/lib/libhpx.so.0
0x2aff4af2ba0c  : ??? + 0x2aff4af2ba0c in /worka/work/parsa/gcc49/hpx-relwithdebinfo/lib/libhpx.so.0
0x2aff4af2cf3d  : hpx::detail::run_or_start(hpx::util::function<int (boost::program_options::variables_map&), void, void> const&, boost::program_options::options_description const&, int, char**, std::vector<std::string, std::allocator<std::string> > const&, hpx::util::function<void (), void, void> const&, hpx::util::function<void (), void, void> const&, hpx::runtime_mode, bool) + 0x29d in /worka/work/parsa/gcc49/hpx-relwithdebinfo/lib/libhpx.so.0
0x438b16        : ??? + 0x438b16 in /worka/work/parsa/gcc49/hpx-relwithdebinfo/bin/1d_stencil_8
0x3cb921ed5d    : __libc_start_main + 0xfd in /lib64/libc.so.6
{what}: Segmentation fault
{config}:
  HPX_HAVE_NATIVE_TLS=ON
  HPX_HAVE_STACKTRACES=ON
  HPX_HAVE_COMPRESSION_BZIP2=OFF
  HPX_HAVE_COMPRESSION_SNAPPY=OFF
  HPX_HAVE_COMPRESSION_ZLIB=OFF
  HPX_HAVE_PARCEL_COALESCING=ON
  HPX_PARCELPORT_TCP=OFF
  HPX_PARCELPORT_MPI=ON (MPICH V3.1, MPI V3.0)
  HPX_PARCELPORT_IPC=OFF
  HPX_PARCELPORT_IBVERBS=OFF
  HPX_HAVE_VERIFY_LOCKS=OFF
  HPX_HAVE_HWLOC=ON
  HPX_HAVE_ITTNOTIFY=OFF
  HPX_RUN_MAIN_EVERYWHERE=OFF
  HPX_LIMIT=5
  HPX_PARCEL_MAX_CONNECTIONS=512
  HPX_PARCEL_MAX_CONNECTIONS_PER_LOCALITY=4
  HPX_INITIAL_AGAS_LOCAL_CACHE_SIZE=256
  HPX_AGAS_LOCAL_CACHE_SIZE_PER_THREAD=32
  HPX_MALLOC=custom
  HPX_PREFIX (configured)=/usr/local
  HPX_PREFIX=/worka/work/parsa/gcc49/hpx-relwithdebinfo
{version}: V0.9.11-trunk (AGAS: V3.0), Git: 9fab4ab8e5
{boost}: V1.57.0
{build-type}: release
{date}: Apr 19 2015 18:39:58
{platform}: linux
{compiler}: GNU C++ version 4.9.0
{stdlib}: GNU libstdc++ version 20140422
[smic004:mpi_rank_0][error_sighandler] Caught error: Aborted (signal 6)
{stack-trace}: 11 frames:
0x2b25f43a36f6  : hpx::termination_handler(int) + 0x126 in /worka/work/parsa/gcc49/hpx-relwithdebinfo/lib/libhpx.so.0
0x357460f710    : ??? + 0x357460f710 in /lib64/libpthread.so.0
0x2b25f4684c6f  : hpx::performance_counters::server::raw_counter::~raw_counter() + 0x1f in /worka/work/parsa/gcc49/hpx-relwithdebinfo/lib/libhpx.so.0
0x2b25f47040ea  : hpx::util::reinit_destruct() + 0x4a in /worka/work/parsa/gcc49/hpx-relwithdebinfo/lib/libhpx.so.0
0x2b25f43a66a1  : hpx::runtime::~runtime() + 0x261 in /worka/work/parsa/gcc49/hpx-relwithdebinfo/lib/libhpx.so.0
0x2b25f435bc0f  : hpx::runtime_impl<hpx::threads::policies::local_priority_queue_scheduler<boost::mutex, hpx::threads::policies::lockfree_fifo, hpx::threads::policies::lockfree_fifo, hpx::threads::policies::lockfree_lifo>, hpx::threads::policies::callback_notifier>::~runtime_impl() + 0x14f in /worka/work/parsa/gcc49/hpx-relwithdebinfo/lib/libhpx.so.0
0x2b25f435bd59  : hpx::runtime_impl<hpx::threads::policies::local_priority_queue_scheduler<boost::mutex, hpx::threads::policies::lockfree_fifo, hpx::threads::policies::lockfree_fifo, hpx::threads::policies::lockfree_lifo>, hpx::threads::policies::callback_notifier>::~runtime_impl() + 0x9 in /worka/work/parsa/gcc49/hpx-relwithdebinfo/lib/libhpx.so.0
0x2b25f4382a0c  : ??? + 0x2b25f4382a0c in /worka/work/parsa/gcc49/hpx-relwithdebinfo/lib/libhpx.so.0
0x2b25f4383f3d  : hpx::detail::run_or_start(hpx::util::function<int (boost::program_options::variables_map&), void, void> const&, boost::program_options::options_description const&, int, char**, std::vector<std::string, std::allocator<std::string> > const&, hpx::util::function<void (), void, void> const&, hpx::util::function<void (), void, void> const&, hpx::runtime_mode, bool) + 0x29d in /worka/work/parsa/gcc49/hpx-relwithdebinfo/lib/libhpx.so.0
0x438b16        : ??? + 0x438b16 in /worka/work/parsa/gcc49/hpx-relwithdebinfo/bin/1d_stencil_8
0x3573e1ed5d    : __libc_start_main + 0xfd in /lib64/libc.so.6
{what}: Segmentation fault
{config}:
  HPX_HAVE_NATIVE_TLS=ON
  HPX_HAVE_STACKTRACES=ON
  HPX_HAVE_COMPRESSION_BZIP2=OFF
  HPX_HAVE_COMPRESSION_SNAPPY=OFF
  HPX_HAVE_COMPRESSION_ZLIB=OFF
  HPX_HAVE_PARCEL_COALESCING=ON
  HPX_PARCELPORT_TCP=OFF
  HPX_PARCELPORT_MPI=ON (MPICH V3.1, MPI V3.0)
  HPX_PARCELPORT_IPC=OFF
  HPX_PARCELPORT_IBVERBS=OFF
  HPX_HAVE_VERIFY_LOCKS=OFF
  HPX_HAVE_HWLOC=ON
  HPX_HAVE_ITTNOTIFY=OFF
  HPX_RUN_MAIN_EVERYWHERE=OFF
  HPX_LIMIT=5
  HPX_PARCEL_MAX_CONNECTIONS=512
  HPX_PARCEL_MAX_CONNECTIONS_PER_LOCALITY=4
  HPX_INITIAL_AGAS_LOCAL_CACHE_SIZE=256
  HPX_AGAS_LOCAL_CACHE_SIZE_PER_THREAD=32
  HPX_MALLOC=custom
  HPX_PREFIX (configured)=/usr/local
  HPX_PREFIX=/worka/work/parsa/gcc49/hpx-relwithdebinfo
{version}: V0.9.11-trunk (AGAS: V3.0), Git: 9fab4ab8e5
{boost}: V1.57.0
{build-type}: release
{date}: Apr 19 2015 18:39:58
{platform}: linux
{compiler}: GNU C++ version 4.9.0
{stdlib}: GNU libstdc++ version 20140422
[smic007:mpi_rank_1][error_sighandler] Caught error: Aborted (signal 6)

===================================================================================
=   BAD TERMINATION OF ONE OF YOUR APPLICATION PROCESSES
=   PID 53490 RUNNING AT smic004
=   EXIT CODE: 6
=   CLEANING UP REMAINING PROCESSES
=   YOU CAN IGNORE THE BELOW CLEANUP MESSAGES
===================================================================================
[proxy:0:1@smic007] HYD_pmcd_pmip_control_cmd_cb (pm/pmiserv/pmip_cb.c:913): assert (!closed) failed
[proxy:0:1@smic007] HYDT_dmxu_poll_wait_for_event (tools/demux/demux_poll.c:76): callback returned error status
[proxy:0:1@smic007] main (pm/pmiserv/pmip.c:206): demux engine error waiting for event
[mpiexec@smic004] HYDT_bscu_wait_for_completion (tools/bootstrap/utils/bscu_wait.c:76): one of the processes terminated badly; aborting
[mpiexec@smic004] HYDT_bsci_wait_for_completion (tools/bootstrap/src/bsci_wait.c:23): launcher returned error waiting for completion
[mpiexec@smic004] HYD_pmci_wait_for_completion (pm/pmiserv/pmiserv_pmci.c:218): launcher returned error waiting for completion
[mpiexec@smic004] main (ui/mpich/mpiexec.c:336): process manager error waiting for completion
```

**Comment by sithhell** [comment]
One thing I noticed while trying to reproduce this is that all packages are compiled with the intel compiler, not gcc. Investigating if that cause the issue.

**Comment by sithhell** [comment]
I can't reproduce this with latest master (+fix in the intel_compiler_error branch), intel14 and mvapich.
Here are the steps I did:
```
$ module swap impi/4.1.3.048/intel64 mvapich2/2.0
$ cmake                                                                           \
      -DCMAKE_BUILD_TYPE=RelWithDebInfo                                                \
      -DCMAKE_CXX_COMPILER=icpc                                                 \
      -DBOOST_ROOT=$LHPC_ROOTBOOST                                              \
      -DBoost_NO_SYSTEM_PATHS=On                                                \
      -DHPX_PARCELPORT_MPI=On                                                   \
      -DHPX_PARCELPORT_TCP=Off                                                  \
      -DHPX_PARCELPORT_IBVERBS=Off                                              \
      -DHWLOC_ROOT=$HOME/hwloc/host                                             \
      -DHPX_MALLOC="custom"                                                     \
      -Wdev                                                                     \
      $HOME/hpx/source
$ make -j8 1d_stencil_8_exe
$ qsub -q workq -l nodes=2:ppn=20 -l walltime=1:00:00 -I
$ module swap impi/4.1.3.048/intel64 mvapich2/2.0
$ cd $PBS_O_WORKDIR
$ mpirun_rsh -ssh -np 1 $(head $PBS_NODEFILE) ./hpx/host/relwithdebinfo/bin/1d_stencil_8 --nx 100000 --np 2000 -t 20
```

I tried this several times. Didn't encounter any failure. Could it be that you ran your code with impi instead of mvapich2?

**Comment by parsa** [comment]
>[11:08:53]  <parsa>	 heller: i have no idea what you did.... i don't have impi loaded mvapich2 is the default and that's what i have... #1467 still exists on current master... and i'm not quite sure if you ran what you've posted or at least the line with mpirun_rsh: http://fpaste.org/215893/43015452/

**Comment by sithhell** [comment]
I am just not able to reproduce this. Please observe the gist here, where I recorded each and every command (with relevant output):
https://gist.github.com/sithhell/3cca3430a3969bf1eaca

**Comment by parsa** [comment]
It goes away if both `mvapich2/2.0/INTEL-14.0.2` (default MPI lib on SuperMIC) and `INTEL-140-MVAPICH2` are loaded and the application is run with `mpirun_rsh -ssh -np $PBS_NUM_NODES $(uniq $PBS_NODEFILE) $APP_PATH --nx 100000 --np 2000 -t 20`.

It still does happen with `mpirun` though.

Not sure if we should close this yet.

**Comment by hkaiser** [comment]
As a minimum we should properly document this issue.

**Comment by sithhell** [comment]
I am still surprised that `mpirun` works. What is the output of `which mpirun` in an interactive session?
In any case, this is *not* an HPX issue.

**Comment by parsa** [comment]
```
$ which mpirun
+ /usr/bin/which --tty-only --read-alias --show-dot --show-tilde mpirun
+ alias
/usr/local/packages/mvapich2/2.0/INTEL-14.0.2/bin/mpirun
$ which mpirun_rsh
+ alias
+ /usr/bin/which --tty-only --read-alias --show-dot --show-tilde mpirun_rsh
/usr/local/packages/mvapich2/2.0/INTEL-14.0.2/bin/mpirun_rsh
```
Not convinced this isn't an HPX issue. Happens only with Intel 14 and all libraries used are also compiled by the same compiler. Also the application runs and SegFaults after the actual application is finished. It *is* an HPX issue.

**Comment by sithhell** [comment]
On Wednesday, April 29, 2015 14:18:31 Parsa Amini wrote:
> ```
> $ which mpirun
> + /usr/bin/which --tty-only --read-alias --show-dot --show-tilde mpirun
> + alias
> /usr/local/packages/mvapich2/2.0/INTEL-14.0.2/bin/mpirun
> $ which mpirun_rsh
> + alias
> + /usr/bin/which --tty-only --read-alias --show-dot --show-tilde mpirun_rsh
> /usr/local/packages/mvapich2/2.0/INTEL-14.0.2/bin/mpirun_rsh
> ```
> Not convinced this isn't an HPX issue. Happens only with Intel 14 and all
> libraries used are also compiled by the same compiler. Also the application
> runs and SegFaults after the actual application is finished. It *is* an HPX
> issue.

As said, I can't reproduce any of this behavior in this ticket. Even a simple 
`mpirun hostname` gives me the linker error in the gist above.

**Comment by parsa** [comment]
@sithhell I have had that link error. After I got tech support from Loni it turned out to be of the `if`s in my bashrc. You might want to look into that.

**Comment by hkaiser** [comment]
> Not convinced this isn't an HPX issue. Happens only with Intel 14 and all dependencies are also compiled by the same compiler. Also the application runs and SegFaults after the actual application is finished. It is an HPX issue.

This seems to happen only if there is some module mismatch. Loading the correct module apparently fixes the issue. I'd tend to close this ticket. We should document this subtle problem for SuperMic, however.

**Comment by sithhell** [comment]
Am 01.05.2015 4:02 vorm. schrieb "Hartmut Kaiser" <notifications@github.com
>:
>>
>> Not convinced this isn't an HPX issue. Happens only with Intel 14 and
all dependencies are also compiled by the same compiler. Also the
application runs and SegFaults after the actual application is finished. It
is an HPX issue.
>
> This seems to happen only if there is some module mismatch. Loading the
correct module apparently fixes the issue. I'd tend to close this ticket.
We should document this subtle problem for SuperMic, however.

I still can't reproduce this issue. I eventually figured out the linker
problem, but the segfault is not reproducible for me. Closing.
